dig.go

```diff
@@ -119,6 +143,15 @@ func (c *Container) Provide(constructor interface{}, opts ...ProvideOption) erro
 // The function may return an error to indicate failure. The error will be
 // returned to the caller as-is.
 func (c *Container) Invoke(function interface{}, opts ...InvokeOption) error {
+	err := c.invoke(function, opts...)
+	// If error wrapping wasn't requested, return the root cause.
+	if !c.wrapErrors {
```

Comment:
does `Provide` need the same logic too?

Reply:
No - a constructor returning an error should always result in the program failing in the same way it does currently.

Reply:
@abhinav you should specify on this PR and the issue that this change is intended to target `Invoke` only.

Reply:
I don't see how you can realistically differentiate between one but not the other. If `.New` fails with a constructor error it's going to be all muffled anyways, something like `failed to create type X: blah blah: blah blah`, which may or may not be desirable either.

That said, I think I'd prefer @prashantv 's suggestion of having an extract function. That way the user can make their own decision of what is and isn't desirable.

Reply:
I think @breerly's point was that functions don't get called on Provide so that path remains unchanged. Functions get called only on the Invoke path, so that's the only place where we needed to consider this option.

Reply:
Yeah, it makes sense - we want to bubble up user constructor errors back up the stack through `RootCause`.

I was just saying that the name `RootCause` in relation to just `Invoke` can be pretty confusing. Technically, `.New` errors when provides fail also have a "root cause": bad argument types, no returns, etc. etc.